src/main/java/org/opentripplanner/routing/edgetype/BikeRentalEdge.java

```diff
@@ -20,65 +19,66 @@
 
     private static final long serialVersionUID = 1L;
 
     private Set<String> networks;
 
-    public RentABikeAbstractEdge(Vertex from, Vertex to, Set<String> networks) {
+    public BikeRentalEdge(Vertex from, Vertex to, Set<String> networks) {
         super(from, to);
         this.networks = networks;
     }
 
-    protected State traverseRent(State s0) {
+    public State traverse(State s0) {
+        if (!s0.getOptions().bikeRental) { return null; }
+
+        StateEditor s1 = s0.edit(this);
         RoutingRequest options = s0.getOptions();
-        /*
-         * If we already have a bike (rented or own) we won't go any faster by having a second one.
-         */
-        if (!s0.getNonTransitMode().equals(TraverseMode.WALK))
-            return null;
-        /*
-         * To rent a bike, we need to have BICYCLE in allowed modes.
-         */
-        if (!options.streetSubRequestModes.contains(TraverseMode.BICYCLE))
-            return null;
 
-        BikeRentalStationVertex dropoff = (BikeRentalStationVertex) tov;
-        if (options.useBikeRentalAvailabilityInformation && dropoff.getBikesAvailable() == 0) {
-            return null;
+        BikeRentalStationVertex stationVertex = (BikeRentalStationVertex) tov;
+
+        if (options.arriveBy) {
+            switch (s0.getBikeRentalState()) {
+                case BEFORE_RENTING:
+                    return null;
+                case HAVE_RENTED:
+                    s1.dropOffRentedVehicleAtStation(stationVertex.getVehicleMode(),true);
+                    break;
+                case RENTING_FLOATING:
+                    s1.beginFloatingVehicleRenting(stationVertex.getVehicleMode(), true);
+                case RENTING_FROM_STATION:
+                    if (!hasCompatibleNetworks(networks, s0.getBikeRentalNetworks())) { return null; }
+                    s1.beginVehicleRentingAtStation(stationVertex.getVehicleMode(), true);
+                    break;
+            }
+        } else {
+            switch (s0.getBikeRentalState()) {
+                case BEFORE_RENTING:
+                    if (options.useBikeRentalAvailabilityInformation && stationVertex.getBikesAvailable() == 0) {
+                        return null;
+                    }
+                    if (stationVertex.getStation().isFloatingBike) {
+                        s1.beginFloatingVehicleRenting(stationVertex.getVehicleMode(), false);
+                    } else {
+                        s1.beginVehicleRentingAtStation(stationVertex.getVehicleMode(), false);
+                    }
+                    break;
+                case HAVE_RENTED:
+                case RENTING_FLOATING:
+                    return null;
+                case RENTING_FROM_STATION:
+                    if (!hasCompatibleNetworks(networks, s0.getBikeRentalNetworks())) { return null; }
+                    s1.dropOffRentedVehicleAtStation(stationVertex.getVehicleMode(), false);
+                    break;
```

Comment:
This does not appear to allow for the use case of a floating vehicle having been rented and then being returned at a station. This modification should remedy that situation:

```suggestion
                case HAVE_RENTED:
                    return null;
                case RENTING_FLOATING:
                case RENTING_FROM_STATION:
                    if (!hasCompatibleNetworks(networks, s0.getBikeRentalNetworks())) { return null; }
                    s1.dropOffRentedVehicleAtStation(stationVertex.getVehicleMode(), false);
                    break;
```

One other thing to point out is that there isn't a check for whether a vehicle is being returned to the station that it was picked up from. Such situations should probably not be allowed.

Reply:
One more thing I thought of is that there should be a check for whether this station is a floating bike. In that case dropoffs should not be allowed.